h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java

```diff
@@ -1218,23 +1218,44 @@ public void testSparsityDetection(){
     XGBoostModel sparseModel = null;
     XGBoostModel denseModel = null;
     try {
-      Frame sparseFrame = Scope.track(TestUtil.generate_enum_only(2, 10, 10, 0));
-      Frame denseFrame = Scope.track(TestUtil.generate_enum_only(2, 10, 2, 0));
+
+      // Fill ratio 0.2 (response not counted)
+      Frame sparseFrame = Scope.track(new TestFrameBuilder()
+              .withName("sparse_frame")
+              .withColNames("C1", "C2", "C3")
+              .withVecTypes(Vec.T_NUM,Vec.T_NUM,Vec.T_NUM)
+              .withDataForCol(0, ard( 1,0, 0, 0, 0))
+              .withDataForCol(1, ard( 0, 1, 0, 0, 0))
+              .withDataForCol(2, ard( 2, 1,1, 4, 3))
+              .build());
+
 
       XGBoostModel.XGBoostParameters parms = new XGBoostModel.XGBoostParameters();
       parms._train = sparseFrame._key;
-      parms._response_column = "C1";
+      parms._response_column = "C3";
       parms._seed = 42;
+      parms._backend = XGBoostModel.XGBoostParameters.Backend.cpu;
       parms._ntrees = 1;
       parms._dmatrix_type = XGBoostModel.XGBoostParameters.DMatrixType.auto;
 
       sparseModel = (XGBoostModel) Scope.track_generic(new hex.tree.xgboost.XGBoost(parms).trainModel().get());
       assertNotNull(sparseModel);
       assertTrue(sparseModel._output._sparse);
 
+      // Fill ratio 0.3 (response not counted) - the threshold for sprase is >= 0.3
+      Frame denseFrame = Scope.track(new TestFrameBuilder()
```

Comment:
This is really weird to me. There are exactly 15 values in the Frame. Response is not counted when sparsity is calculated, thus only C1 and C1 (10 values in total) are counted.
Only 3 values are non-zero. The `fill ratio` is 0.3. Yet we consider this matrix sprase.

I wonder with current logic in `isTrainDatasetSparse` in `XGBoost` class:

```java
    /**
     * @return True if train dataset is sparse, otherwise false.
     */
    private boolean isTrainDatasetSparse() {
      long nonZeroCount = 0;
      int nonCategoricalColumns = 0;
      long oneHotEncodedColumns = 0;
      for (int i = 0; i < _train.numCols(); ++i) {
        if (_train.name(i).equals(_parms._response_column)) continue;
        if (_train.name(i).equals(_parms._weights_column)) continue;
        if (_train.name(i).equals(_parms._fold_column)) continue;
        if (_train.name(i).equals(_parms._offset_column)) continue;
        final Vec vector = _train.vec(i);
        if (vector.isCategorical()) {
          nonZeroCount += _train.numRows();
        } else {
          nonZeroCount += vector.nzCnt();
        }
        if (vector.isCategorical()) {
          oneHotEncodedColumns += vector.cardinality();
        } else {
          nonCategoricalColumns++;
        }
      }
      final long totalColumns = oneHotEncodedColumns + nonCategoricalColumns;
      final double denominator = (double) totalColumns * _train.numRows();
      final double fillRatio = (double) nonZeroCount / denominator;
      Log.info("fill ratio: " + fillRatio);

      return fillRatio < FILL_RATIO_THRESHOLD
          || ((_train.numRows() * totalColumns) > Integer.MAX_VALUE);
    }
```

shouldn't the `FILL_RATIO_THRESHOLD` be `0.7` instead of current value `0.3` ?